File: backend/app/schemas/friendship.py
from pydantic import BaseModel
from typing import Literal
from app.schemas.search import UserSearchResult


class FriendRequestCreate(BaseModel):
    receiver_id: int


class FriendshipResponse(BaseModel):
    friendship_id: int
    status: Literal["pending", "accepted", "rejected"]
    friend: UserSearchResult
    is_requester: bool

    class Config:
        from_attributes = True


class FriendshipStatusUpdate(BaseModel):
    status: Literal["accepted", "rejected"]


class FriendshipActionResponse(FriendshipResponse):
    message: str


class FriendshipNotFoundError(Exception):
    pass


class FriendshipNotAuthorizedError(Exception):
    pass

File: frontend/src/types/friendship.ts
export interface Friendship {
  friendship_id: number;
  status: "pending" | "accepted" | "rejected";
  friend: UserSearchResult;
  is_requester: boolean;
}

export interface FriendshipAction extends Friendship {
  message: string;
}

export interface UserSearchResult {
  id: number;
  username: string;
  avatar: string;
}

export type FriendshipStatus = "pending" | "accepted" | "rejected";

export type FriendshipStatusToDisplay = "none" | "pending" | "friends";
